bot/utils/redis_dict.py

```diff
@@ -0,0 +1,136 @@
+
+    _namespaces = []
+    _redis = redis_py.Redis(
+        host=constants.Bot.redis_host,
+        port=constants.Bot.redis_port,
+    )  # Can be overridden for testing
+
+    def __init__(self, namespace: Optional[str] = None) -> None:
+        """Initialize the RedisDict with the right namespace."""
+        super().__init__()
+        self._has_custom_namespace = namespace is not None
+
+        if self._has_custom_namespace:
+            self._set_namespace(namespace)
+        else:
+            self.namespace = "global"
+
+    def _set_namespace(self, namespace: str) -> None:
+        """Try to set the namespace, but do not permit collisions."""
+        while namespace in self._namespaces:
+            namespace = namespace + "_"
+
+        self._namespaces.append(namespace)
+        self._namespace = namespace
+
+    def __set_name__(self, owner: object, attribute_name: str) -> None:
+        """
+        Set the namespace to Class.attribute_name.
+
+        Called automatically when this class is constructed inside a class as an attribute, as long as
+        no custom namespace is provided to the constructor.
+        """
+        if not self._has_custom_namespace:
+            self._set_namespace(f"{owner.__name__}.{attribute_name}")
+
+    def __repr__(self) -> str:
+        """Return a beautiful representation of this object instance."""
+        return f"RedisDict(namespace={self._namespace!r})"
+
+    def __eq__(self, other: RedisDict) -> bool:
+        """Check equality between two RedisDicts."""
+        return self.items() == other.items() and self._namespace == other._namespace
+
+    def __ne__(self, other: RedisDict) -> bool:
+        """Check inequality between two RedisDicts."""
+        return self.items() != other.items() or self._namespace != other._namespace
+
+    def __setitem__(self, key: ValidRedisKey, value: JSONSerializableType):
+        """Store an item in the Redis cache."""
+        # JSON serialize the value before storing it.
+        json_value = json.dumps(value)
+        self._redis.hset(self._namespace, key, json_value)
+
+    def __getitem__(self, key: ValidRedisKey):
+        """Get an item from the Redis cache."""
+        value = self._redis.hget(self._namespace, key)
+
+        if value:
+            return json.loads(value)
+
+    def __delitem__(self, key: ValidRedisKey):
+        """Delete an item from the Redis cache."""
+        self._redis.hdel(self._namespace, key)
+
+    def __contains__(self, key: ValidRedisKey):
+        """Check if a key exists in the Redis cache."""
+        return self._redis.hexists(self._namespace, key)
+
+    def __iter__(self):
+        """Iterate all the items in the Redis cache."""
+        keys = self._redis.hkeys(self._namespace)
+        return iter([key.decode('utf-8') for key in keys])
+
+    def __len__(self):
+        """Return the number of items in the Redis cache."""
+        return self._redis.hlen(self._namespace)
+
+    def copy(self) -> Dict:
```

Comment:
I probably would have named this `to_dict`. It feels like the name, `copy`, and the dosctring, `Convert to dict and return`, tell a different story.

Reply:
right, maybe, but this is just an overloaded standard dict method. we're replicating the functionality from `dict.copy()`, which would return a shallow copy of the dict object. We cannot really return a "copy" of the `RedisDict`, so we do the next best thing and just return a `dict`.

`to_dict` is probably a better name, but there's no `dict.to_dict()` - I mean, that would be silly.

Reply:
> this is just an overloaded standard dict method

Right, I didn't realize that. That does justify it a decent amount.